src/core/derivation.ts

```diff
@@ -24,6 +29,30 @@ export interface IDerivation extends IDepTreeNode, IObservable, ISetEntry {
 	 * amount of dependencies used by the derivation in this run, which has not been bound yet.
 	 */
 	unboundDepsCount: number;
+	onBecomeStale();
+	recoverFromError();
+}
+
+export function shouldCompute(derivation: IDerivation, onError?): boolean {
+	const dependenciesState = derivation.dependenciesState;
+	if (dependenciesState === 0) return false;
+	if (dependenciesState === -1 || dependenciesState === 2) return true;
+	// derivation.dependencyChange === 1
+	let hasError = true;
+	try {
+		for (let i = 0; i < derivation.observing.length; i++) {
+			const obj = derivation.observing[i];
+			if (obj instanceof ComputedValue) {
+				obj.get();
+				if (derivation.dependenciesState === 2) return true;
```

Comment:
yep, gonna comment it.
`state === 1` is possible only on `ComputedValue`
at the end of `shouldCompute` `derivation.dependenciesState !== 1`
`ComputedValue.get` will always call shouldCompute